Add sysroot.bootloader repo config key (fix running bootloader when no deployments exist)

rfairley opened the pull request:
The sysroot.bootloader key is used to configure the bootloader
that OSTree uses when deploying a sysroot. Having this key
allows specifying behavior not to use the default bootloader
backend code, which is preferable when creating a first
deployment from the sysroot (#1774).

As of now, the key can take the values "auto" or "none". If
the key is not given, the value defaults to "auto".

"auto" causes _ostree_sysroot_query_bootloader() to be used
when writing a new deployment, which is the original behavior
that dynamically detects which bootloader to use.

"none" avoids querying the bootloader dynamically. The BLS
config fragments are still written to
sysroot/boot/loader/entries for use by higher-level software.

More values can be supported in future to specify a single
bootloader, different behavior for the bootloader code, or
a list of bootloaders to try.

Resolves: #1774

cgwalters commented:
Related: https://github.com/ostreedev/ostree/issues/1719

I need to dive into this more; not entirely sure it's right (but it doesn't seem necessarily wrong either).  There's risk here in changing the grub backend.

I know it's more work but if we can get FCOS systems into having a static grub config (i.e. ostree only writes BLS snippets, use the grub BLS plugin) the whole architecture is much clearer.

rfairley commented:
Thanks @cgwalters - just pushed a change. `new_deployments` seems to hold the pending deployment as the first entry - so I think this is the data that needs to be passed in order for grub2-mkconfig to use the pending deployment.

> There's risk here in changing the grub backend.

Agreed on the risk - I think now this is a question of "could this change break existing systems that might have been relying on the existing deployment being used rather than the pending".

I'm interested in taking a look at using the BLS plugin in OSTree - I need to read into it more, but now that I've had a look at the bootloader code a bit better, I think I can see what would need to be changed.

rfairley commented:
As part of this, going to start on an implementation for writing the BLS config, and a way to configure the bootloader (something like https://github.com/ostreedev/ostree/issues/1801#issuecomment-453854856).

One way after brief discussion is to have OSTree always write BLS configs, and ship with a config `bootloader=auto` to use the auto-detecting bootloader functionality currently in place. This can be configured to `bootloader=none` to **not** use the auto-detecting functionality (only BLS configs will be written).

cgwalters commented:
> One way after brief discussion is to have OSTree always write BLS configs

That's what ostree has done since the very start.  The BLS configs are canonical.  But at the time RHEL7's grub didn't speak BLS (actually no bootloaders did really) so I added the concept of a "bootloader backend" to ostree as a temporary hack.

libostree never *reads* the grub config - it reads the BLS configs it writes as the source of truth.

rfairley commented:
> That's what ostree has done since the very start

Ahh that's clear to me now, thanks!

```
# ls /boot/loader/entries/ostree-
ostree-1-fedora-atomic.conf  ostree-2-fedora-atomic.conf
```

So what's needed is really having a switch in the bootloader backend to use or not use it (at least for the grub2-mkconfig part), and a way to configure that switch (through `bootloader=`)?

rfairley commented:
Dropped the edits to the grub2 bootloader code, and started on adding a `sysroot.bootloader` key to the ostree repo config.


WIP, with the following left to do:
- [ ] Errors are given due to the `[sysroot]` group not existing in the default `repo/config` files. Looking for a way to make that field optional.
- [ ] Update the docs.
- [ ] Add a test.

Right now I'm testing with the following commands:

```
mkdir sysroot && ostree admin init-fs sysroot
echo "[sysroot]" >> sysroot/ostree/repo/config
echo "bootloader=none" >> sysroot/ostree/repo/config
ostree pull-local --repo=sysroot/ostree/repo /sysroot/ostree/repo
ostree admin os-init osname --sysroot sysroot
ref=$(ostree refs --repo sysroot/ostree/repo) && mkdir -p sysroot/boot/grub2 && touch sysroot/boot/grub2/grub.cfg
ostree admin deploy "$ref" --sysroot sysroot --os osname -v 2>err.txt
ls sysroot/boot/loader/entries/ostree-1-osname.conf
```

jlebon commented on src/libostree/ostree-repo.c:
```diff
@@ -3113,6 +3113,53 @@ reload_remote_config (OstreeRepo          *self,
+      {
+        /* XXX: quick hack to make existing config files not fail
+         * on not having the [sysroot] group present. */
+        if (error)
```
The pattern for this normally is creating a local_error so you can always check for the exact error, and then `g_propagate_error` if needed.

rfairley replied:
Got it - makes sense.

jlebon commented on src/libostree/ostree-repo.c:
```diff
@@ -3113,6 +3113,53 @@ reload_remote_config (OstreeRepo          *self,
+                       GCancellable        *cancellable,
+                       GError             **error)
+{
+  { g_autofree char *bootloader = NULL;
```
Why the nested scope here?

jlebon replied:
Ahh I see, in case we have more configs in `[sysroot]`, I guess?

rfairley replied:
Yes - that was it. I was mostly copying the pattern from the `reload_core_config()` and `reload_remtoe_config()` - but also in case we add more to `[sysroot]`.

jlebon commented on src/libostree/ostree-repo.c:
```diff
@@ -3113,6 +3113,53 @@ reload_remote_config (OstreeRepo          *self,
+     * https://github.com/ostreedev/ostree/issues/1801
+     */
+    if (strcmp (bootloader, "auto") != 0 &&
+        strcmp (bootloader, "none") != 0)
```
`!g_str_equal`

rfairley replied:
Fixing!

jlebon commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -2310,6 +2311,7 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
   gboolean bootloader_is_atomic = FALSE;
   SyncStats syncstats = { 0, };
   g_autoptr(OstreeBootloader) bootloader = NULL;
+  g_autofree const char *bootloader_config = NULL;
```
That should just be `const char*`, no?

rfairley replied:
Ahh, yes, shouldn't be freeing that here.

rfairley commented:
I'm thinking bd082adaca9a7e95207341af8e7e1e832896f0fc is a clean way to make the `sysroot` group optional by extending the existing API to give a default value if the group is not found.

So far, the `core` group uses this API, aside from a case where we don't handle the error: https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-repo.c#L2928.

A problem I see with this is not giving the user an error if for some reason `[core]` is removed from the config file. Config files where the key exists but the group `[core]` does not would still have default values applied (wherever `ot_keyfile_get_value_with_default` is used). This shouldn't be a problem if the config is done through `ostree config set` though, and I think the documentation showing `[core]` being needed in the file is sufficient.

rfairley commented:
Hmm - seeing the tests for keyfile utils https://github.com/ostreedev/ostree/blob/master/tests/test-keyfile-utils.c#L126, probably best to keep the API as it is and instead only make the `[sysroot]` section optional (by checking if the group exists in `reload_sysroot_config` and just returning success if it doesn't exist).

rfairley commented:
Added tests for the `none` bootloader config option. Ready for another round of review / comments on the approach.

For unit tests for the added keyfile util function `ot_keyfile_get_value_with_default_group_optional()`, the only change from `ot_keyfile_get_value_with_default()` is here: https://github.com/ostreedev/ostree/pull/1814/commits/1f181097da8a1070cc2f6b45e6b6ba0c97dcdab1#diff-11958a5431bcbe882db3033c8de63f5eR190.

Also updated the first comment of this PR to match the commit description of  [`71915df`](/ostreedev/ostree/commit/71915df5e148c4593ba3d44d1f1ebfd16d79c168).

rfairley commented:
Noticed this is still failing the CI tests on `tests/test-admin-deploy-none.sh`. Will fix these, and add a note in docs for the new option.

rfairley commented:
Think that should fix the failing test, `${CMD_PREFIX}` was not present on one of the `ostree` commands.

Added some documentation in the ostree repo config manpage - lifting WIP

rfairley commented:
The reason `test-admin-deploy-none.sh` was failing ([log](https://s3.amazonaws.com/aos-ci/ghprb/ostreedev/ostree/d37852775ae20ef47ddc821c04d75ddc1c2c1740.0.1550846924962181195/artifacts/gdtr-results/libostree_test-admin-deploy-none.sh.test.txt)) with error  `fsetxattr: Operation not supported` was due to setting xattrs not supported on tmpfs (same as https://github.com/projectatomic/atomic/issues/1247). This occurred during `admin deploy` while copying `/etc/a-new-default-config-file` which had been committed to the tree in `admin-test.sh`. Updated the test to not require copying that file, by setting up a clean repo with `setup_os_repo()` and deploying the tree from there.

Just tidied up the test, a bit - should be passing now.

cgwalters commented on man/ostree.repo-config.xml:
```diff
@@ -333,6 +333,41 @@ Boston, MA 02111-1307, USA.
   </refsect1>
 
+  <refsect1>
+    <title>[sysrooot] Section Options</title>
```
`sysroot`

cgwalters commented on man/ostree.repo-config.xml:
```diff
@@ -333,6 +333,41 @@ Boston, MA 02111-1307, USA.
+    <title>[sysrooot] Section Options</title>
+
+    <para>
+      Options for the sysroot, which OSTree deploys the OSTree repository to.  The following entries are defined:
```
How about "which contains the OSTree repository, deployments, and stateroots".

rfairley replied:
:+1: SGTM, thanks

cgwalters commented on man/ostree.repo-config.xml:
```diff
@@ -333,6 +333,41 @@ Boston, MA 02111-1307, USA.
+          for the deployment.
+        </para>
+        <para>
+          If <literal>auto</literal>, then OSTree will dynamically check for the
```
"then in addition to generating the BLS fragments, OSTree will..."

cgwalters commented on src/libostree/ostree-repo.c:
```diff
@@ -3113,6 +3113,27 @@ reload_remote_config (OstreeRepo          *self,
+      return FALSE;
+
+    if (bootloader == NULL)
+      return glnx_throw (error, "Invalid empty bootloader configuration");
```
I don't think this case can happen...did you mean to compare vs the empty string `""`?  Either way it should probably be instead something like:

```
if (!(g_str_equal (bootloader, "auto") || g_str_equal (bootloader, "none"))
  return glnx_throw (error, "Invalid bootloader configuration: '%s'", bootloader);
```

rfairley replied:
Hmm, for some reason thought it would occur with `bootloader=` in the keyfile, but looks like this just returns an empty string.

Initially I was thinking that `reload_sysroot_config()` should allow any value for `bootloader=$value`, and then have `ostree_sysroot_write_deployments_with_options()` search for something that matches the value (e.g. if we later supported a `/etc/ostree/bootloaders`, `ostree_sysroot_write_deployments_with_options()` would search for `/etc/ostree/bootloaders/my_bootloader` for `bootloader=my_bootloader`). I agree now that validating right at the parsing for this change is better - can always be reworked if `/etc/ostree/bootloaders` or `/usr/lib/ostree/bootloaders` is implemented. Fixing!

cgwalters commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -2342,8 +2343,36 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
+        {
+          /* No bootloader specified; do not query bootloaders to run. */
+        }
+      else
```
Then we can drop the code here.  Seems better to me to validate it right after parsing above.

rfairley replied:
:+1: Fixing

rfairley commented:
Working on the review fixes - will try fixing #1827 so that a test

```
if ${CMD_PREFIX} ostree --repo=sysroot/ostree/repo config set sysroot.bootloader some_unsupported_bootloader 2>err.txt; then
    assert_not_reached "unexpectedly succeeded at setting an unsupportd bootloader config"
fi
assert_file_has_content err.txt "Invalid bootloader configuration.*some_unsupported_bootloader"
echo "ok fail on unsupported bootloader config"
```

can be added (right now unexpectedly succeeds).

rfairley commented:
^ Pushed fixes for the review comments. Since a test case is present for the functionality in this PR, and the failing config test cases are likely to change soon if additional generic bootloader config options are added, I'm thinking it's OK to defer on those tests, and add them in if #1827 is fixed soon.

rfairley commented:
Sorry for the multiple pushes - the diff from the last 4 pushes (in response to review): https://github.com/ostreedev/ostree/compare/27c643a062844e24e79819c2b24bcb51c37b76cd..1ee554d05a506d18eccbea0d69bf356be15fba9d

rfairley commented on src/libostree/ostree-repo.c:
```diff
@@ -5186,8 +5186,12 @@ _ostree_repo_gpg_verify_data_internal (OstreeRepo    *self,
 
```
Thanks, fixed!

rfairley commented on src/libotutil/ot-keyfile-utils.c:
```diff
@@ -101,6 +101,44 @@ ot_keyfile_get_value_with_default (GKeyFile      *keyfile,
+                                                  const char    *default_value,
+                                                  char         **out_value,
+                                                  GError       **error)
+{
```
:+1: sounds good to me, made this change. I was stuck deciding between the two before (calling out to `ot_keyfile_get_value_with-default` vs re-implementing/copy+paste). Seems better overall to call out and avoid the duplicate code.

rfairley commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -2342,8 +2343,24 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
-      if (!_ostree_sysroot_query_bootloader (self, &bootloader, cancellable, error))
-        return FALSE;
+      g_autoptr(OstreeRepo) repo = NULL;
+      if (!ostree_sysroot_get_repo (self, &repo, cancellable, error))
```
Fixed!

rfairley commented:
Updated, and fixed the commit messages!
Diff from last review: https://github.com/ostreedev/ostree/compare/1ee554d05a506d18eccbea0d69bf356be15fba9d..2c3cd83a913d395b500b4528f78073838a2d43ab

rfairley commented:
Also split https://github.com/ostreedev/ostree/pull/1814/commits/5b5093f90318433192f7ff7b1e2c59ed8eb7d534 into its own commit (previously added tests for `sysroot.bootloader` there and did this edit in passing, but later removed the tests).

jlebon commented:
Awesome, LGTM!
@rh-atomic-bot r+ 2c3cd83

jlebon commented:
@rh-atomic-bot retry

dustymabe commented:
❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️

AdrianVovk commented:
:+1:

AdrianVovk commented:
:+1: Great work!

rfairley commented:
Thanks :)
